src/internal/stream.settings.cpp

```diff
@@ -0,0 +1,539 @@
+                               trimmed.rend(),
+                               [](char c) { return !std::isspace(c); })
+                    .base(),
+                  trimmed.end());
+
+    return trimmed;
+}
+} // namespace
+
+/* Create and destroy */
+ZarrStreamSettings*
+ZarrStreamSettings_create()
+{
+    try {
+        return new ZarrStreamSettings();
+    } catch (const std::bad_alloc&) {
+        return nullptr;
+    }
+}
+
+void
+ZarrStreamSettings_destroy(ZarrStreamSettings* settings)
+{
+    delete settings;
+}
+
+ZarrStreamSettings*
+ZarrStreamSettings_copy(const ZarrStreamSettings* settings)
+{
+    if (!settings) {
+        LOG_ERROR("Null pointer: settings");
+        return nullptr;
+    }
+
+    ZarrStreamSettings* copy = ZarrStreamSettings_create();
+    if (!copy) {
+        LOG_ERROR("Failed to allocate memory for copy");
+        return nullptr;
+    }
+
+    copy->store_path = settings->store_path;
+    copy->s3_endpoint = settings->s3_endpoint;
+    copy->s3_bucket_name = settings->s3_bucket_name;
+    copy->s3_access_key_id = settings->s3_access_key_id;
+    copy->s3_secret_access_key = settings->s3_secret_access_key;
+    copy->external_metadata = settings->external_metadata;
+    copy->dtype = settings->dtype;
+    copy->compressor = settings->compressor;
+    copy->compression_codec = settings->compression_codec;
+    copy->compression_level = settings->compression_level;
+    copy->compression_shuffle = settings->compression_shuffle;
+    copy->dimensions = settings->dimensions;
+    copy->multiscale = settings->multiscale;
+
+    return copy;
+}
+
+/* Setters */
+ZarrError
+ZarrStreamSettings_set_store_path(ZarrStreamSettings* settings,
+                                  const char* store_path,
+                                  size_t bytes_of_store_path)
+{
+    SETTINGS_SET_STRING(settings, store_path, bytes_of_store_path);
+}
+
+ZarrError
+ZarrStreamSettings_set_s3_endpoint(ZarrStreamSettings* settings,
+                                   const char* s3_endpoint,
+                                   size_t bytes_of_s3_endpoint)
+{
+    SETTINGS_SET_STRING(settings, s3_endpoint, bytes_of_s3_endpoint);
+}
+
+ZarrError
+ZarrStreamSettings_set_s3_bucket_name(ZarrStreamSettings* settings,
+                                      const char* s3_bucket_name,
+                                      size_t bytes_of_s3_bucket_name)
+{
+    SETTINGS_SET_STRING(settings, s3_bucket_name, bytes_of_s3_bucket_name);
+}
+
+ZarrError
+ZarrStreamSettings_set_s3_access_key_id(ZarrStreamSettings* settings,
+                                        const char* s3_access_key_id,
+                                        size_t bytes_of_s3_access_key_id)
+{
+    SETTINGS_SET_STRING(settings, s3_access_key_id, bytes_of_s3_access_key_id);
+}
+
+ZarrError
+ZarrStreamSettings_set_s3_secret_access_key(
+  ZarrStreamSettings* settings,
+  const char* s3_secret_access_key,
+  size_t bytes_of_s3_secret_access_key)
```

Comment:
Nit: Are we using clang-tidy for code formatting? If we don't, we should probably configure it for consistency.